packages/driver/src/cy/assertions.js

```diff
@@ -0,0 +1,509 @@
+    obj.subject = value
+
+    if (_.isUndefined(actual) || (actual !== expected)) {
+      delete obj.actual
+      delete obj.expected
+    }
+  }
+
+  return obj
+}
+
+const create = function (state, queue, retryFn) {
+  const getUpcomingAssertions = () => {
+    const index = state('index') + 1
+
+    const assertions = []
+
+    // grab the rest of the queue'd commands
+    for (let cmd of queue.slice(index).get()) {
+      // don't break on utilities, just skip over them
+      if (cmd.is('utility')) {
+        continue
+      }
+
+      // grab all of the queued commands which are
+      // assertions and match our current chainerId
+      if (cmd.is('assertion')) {
+        assertions.push(cmd)
+      } else {
+        break
+      }
+    }
+
+    return assertions
+  }
+
+  const injectAssertionFns = (cmds) => {
+    return _.map(cmds, injectAssertion)
+  }
+
+  const injectAssertion = (cmd) => {
+    return ((subject) => {
+      // set assertions to itself or empty array
+      if (!cmd.get('assertions')) {
+        cmd.set('assertions', [])
+      }
+
+      // reset the assertion index back to 0
+      // so we can track assertions and merge
+      // them up with existing ones
+      cmd.set('assertionIndex', 0)
+
+      return cmd.get('fn').originalFn.apply(
+        state('ctx'),
+        [subject].concat(cmd.get('args'))
+      )
+    })
+  }
+
+  const finishAssertions = (assertions) => {
+    return _.each(assertions, (log) => {
+      log.snapshot()
+
+      const e = log.get('_error')
+
+      if (e) {
+        return log.error(e)
+      }
+
+      return log.end()
+    })
+  }
+
+  const verifyUpcomingAssertions = function (subject, options = {}, callbacks = {}) {
+    const cmds = getUpcomingAssertions()
+
+    state('upcomingAssertions', cmds)
+
+    // we're applying the default assertion in the
+    // case where there are no upcoming assertion commands
+    const isDefaultAssertionErr = cmds.length === 0
+
+    if (options.assertions == null) {
+      options.assertions = []
+    }
+
+    _.defaults(callbacks, {
+      ensureExistenceFor: 'dom',
+    })
+
+    const ensureExistence = function () {
+      // by default, ensure existence for dom subjects,
+      // but not non-dom subjects
+      switch (callbacks.ensureExistenceFor) {
+        case 'dom': {
+          const $el = determineEl(options.$el, subject)
+
+          if (!$dom.isJquery($el)) {
+            return
+          }
+
+          return cy.ensureElExistence($el)
+        }
+        case 'subject':
+          return cy.ensureExistence(subject)
+
+        default:
+          return
+      }
+    }
+
+    const determineEl = function ($el, subject) {
+      // prefer $el unless it is strickly undefined
+      if (!_.isUndefined($el)) {
+        return $el
+      }
+
+      return subject
+    }
+
+    const onPassFn = () => {
+      if (_.isFunction(callbacks.onPass)) {
+        return callbacks.onPass.call(this, cmds, options.assertions)
```

Comment:
Is this calling the correct `this` with the `=>`?

Reply:
In CoffeeScript code, it was an arrow function like this:

```coffee
onPassFn = =>
  if _.isFunction(callbacks.onPass)
    callbacks.onPass.call(@, cmds, options.assertions)
  else
    subject
```